tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15419,6 +15459,14 @@ void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD,
   return;
 }
 
+// The DiagnoseEntry function does 3 things:
+// 1. Attempt to add the target profile as an implicit shader attribute which is
+// equivalent to determining whether or not this function is the current active
+// entry point or not.
+// 2. Diagnose whether or not all entry point attributes on
+// the decl belong on the decl given the specific target profile being used.
+// 3. Given the specific target profile, verifying that all necessary attributes
+// that must accompany that target profile exist as attributes to the decl.
```

Comment:
Point 1 doesn't sound quite right to me: "which is equivalent to ..."
Point 2 as well: "given the specific target profile being used"
Point 3 is too attribute centric.  These should be verifying any constraints we can now that we have the full entry function decl, including both parameters and attributes.

Here's an attempt at a rewrite:

Determine whether this function is the current entry point for a non-library compilation, add an implicit shader attribute if so.
For an entry point function, now identified by the shader attribute, diagnose entry point constraints:
- diagnose whether entry point attributes are allowed on the entry point type (ShaderKind) at all.
- diagnose the full entry point decl for required attributes, constraints on or between attributes and parameters, and more.